File: src/ManagementPlatform.Application/UseCases/Checkouts/CheckoutService.cs
using System.Text.Json;
using ManagementPlatform.Domain;

namespace ManagementPlatform.Application;

public sealed class CheckoutService(
    IOrderRepository orderRepository,
    ICheckoutRepository checkoutRepository,
    IAppDbSession appDbSession,
    IPaymentGateway paymentGateway,
    IClock clock,
    PaymentRetryOptions paymentRetryOptions)
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public async Task<CheckoutResponse> CheckoutAsync(
        long orderId,
        CheckoutRequest request,
        CancellationToken cancellationToken)
    {
        var idempotencyKey = ValidateAndNormalizeRequest(request);

        var existingAttempt = await checkoutRepository.GetByOrderAndIdempotencyKeyAsync(
            orderId,
            idempotencyKey,
            cancellationToken);

        if (existingAttempt is not null)
        {
            return ToResponse(existingAttempt);
        }

        var order = await LoadValidOrderAsync(orderId, cancellationToken);
        var attempt = await CreatePendingAttemptAsync(order, idempotencyKey, cancellationToken);

        var paymentResult = await ChargeWithRetryAsync(order, attempt, request.PaymentMethodToken, cancellationToken);
        CreateAndAttachPaymentTransaction(order, attempt, paymentResult);

        if (!paymentResult.Succeeded)
        {
            await HandlePaymentFailureAsync(order, attempt, paymentResult, cancellationToken);
            return ToResponse(attempt);
        }

        MarkPaymentSuccess(order, attempt);
        CreateAndAttachInvoice(attempt);
        CreateAndAttachOutboxMessages(order, attempt);

        await appDbSession.SaveChangesAsync(cancellationToken);
        return ToResponse(attempt);
    }

    public async Task<CheckoutResponse> GetAsync(long checkoutId, CancellationToken cancellationToken)
    {
        var attempt = await checkoutRepository.GetByIdAsync(checkoutId, cancellationToken);

        return attempt is null
            ? throw new NotFoundException("Checkout was not found.")
            : ToResponse(attempt);
    }

    private static string ValidateAndNormalizeRequest(CheckoutRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.IdempotencyKey))
        {
            throw new ValidationException("An idempotency key is required.");
        }

        if (string.IsNullOrWhiteSpace(request.PaymentMethodToken))
        {
            throw new ValidationException("A payment method token is required.");
        }

        return request.IdempotencyKey.Trim();
    }

    private async Task<Order> LoadValidOrderAsync(long orderId, CancellationToken cancellationToken)
    {
        var order = await orderRepository.GetForCheckoutAsync(orderId, cancellationToken);

        if (order is null)
        {
            throw new NotFoundException("Order was not found.");
        }

        if (order.Status is not OrderStatus.Draft)
        {
            throw new ConflictException("Order is already paid or being processed.");
        }

        return order;
    }

    private async Task<CheckoutAttempt> CreatePendingAttemptAsync(
        Order order,
        string idempotencyKey,
        CancellationToken cancellationToken)
    {
        var now = clock.UtcNow;
        order.Status = OrderStatus.CheckoutProcessing;

        var attempt = new CheckoutAttempt
        {
            OrderId = order.Id,
            IdempotencyKey = idempotencyKey,
            Status = CheckoutStatus.PaymentPending,
            CreatedAt = now
        };

        checkoutRepository.AddAttempt(attempt);
        await appDbSession.SaveChangesAsync(cancellationToken);

        return attempt;
    }

    private void CreateAndAttachPaymentTransaction(
        Order order,
        CheckoutAttempt attempt,
        RetriedPaymentResult paymentResult)
    {
        var paymentTransaction = new PaymentTransaction
        {
            CheckoutAttemptId = attempt.Id,
            Status = paymentResult.Succeeded ? PaymentStatus.Succeeded : PaymentStatus.Failed,
            AttemptCount = paymentResult.AttemptCount,
            Amount = order.Amount,
            Currency = order.Currency,
            ProviderTransactionId = paymentResult.ProviderTransactionId,
            FailureReason = paymentResult.FailureReason,
            CreatedAt = clock.UtcNow
        };

        attempt.PaymentTransaction = paymentTransaction;
        checkoutRepository.AddPaymentTransaction(paymentTransaction);
    }

    private async Task HandlePaymentFailureAsync(
        Order order,
        CheckoutAttempt attempt,
        RetriedPaymentResult paymentResult,
        CancellationToken cancellationToken)
    {
        attempt.Status = CheckoutStatus.PaymentFailed;
        attempt.FailureReason = paymentResult.FailureReason ?? "Payment was declined.";
        attempt.CompletedAt = clock.UtcNow;
        order.Status = OrderStatus.Draft;

        await appDbSession.SaveChangesAsync(cancellationToken);
    }

    private void MarkPaymentSuccess(Order order, CheckoutAttempt attempt)
    {
        attempt.Status = CheckoutStatus.PaymentSucceeded;
        attempt.CompletedAt = clock.UtcNow;
        order.Status = OrderStatus.Paid;
        order.PaidAt = clock.UtcNow;
    }

    private void CreateAndAttachInvoice(CheckoutAttempt attempt)
    {
        var invoice = new Invoice
        {
            CheckoutAttemptId = attempt.Id,
            Status = InvoiceStatus.Pending,
            CreatedAt = clock.UtcNow
        };

        attempt.Invoice = invoice;
        checkoutRepository.AddInvoice(invoice);
    }

    private void CreateAndAttachOutboxMessages(Order order, CheckoutAttempt attempt)
    {
        var emailPayload = new CheckoutEmailPayload(
            attempt.Id,
            order.Id,
            order.Name,
            order.Tenant.Email,
            order.Amount,
            order.Currency);

        var productionPayload = new ProductionOrderPayload(
            attempt.Id,
            order.Id,
            order.Name,
            order.Amount,
            order.Currency);

        var outboxMessages = new[]
        {
            CreateOutboxMessage(attempt.Id, OutboxMessageType.SendCheckoutEmail, emailPayload),
            CreateOutboxMessage(attempt.Id, OutboxMessageType.PushToProduction, productionPayload)
        };

        attempt.OutboxMessages.AddRange(outboxMessages);
        checkoutRepository.AddOutboxMessages(outboxMessages);
    }

    private async Task<RetriedPaymentResult> ChargeWithRetryAsync(
        Order order,
        CheckoutAttempt attempt,
        string paymentMethodToken,
        CancellationToken cancellationToken)
    {
        var maxAttempts = Math.Max(paymentRetryOptions.MaxAttempts, 1);
        PaymentGatewayResult? latestResult = null;

        for (var attemptNumber = 1; attemptNumber <= maxAttempts; attemptNumber++)
        {
            latestResult = await paymentGateway.ChargeAsync(
                new PaymentGatewayRequest(
                    order.Id,
                    attempt.Id,
                    order.Amount,
                    order.Currency,
                    paymentMethodToken),
                cancellationToken);

            if (latestResult.Succeeded)
            {
                return new RetriedPaymentResult(
                    true,
                    latestResult.ProviderTransactionId,
                    null,
                    attemptNumber);
            }

            if (!latestResult.IsRetryable || attemptNumber == maxAttempts)
            {
                return new RetriedPaymentResult(
                    false,
                    null,
                    latestResult.FailureReason,
                    attemptNumber);
            }

            var delay = Math.Max(paymentRetryOptions.DelayMilliseconds, 0);
            if (delay > 0)
            {
                await Task.Delay(delay, cancellationToken);
            }
        }

        return new RetriedPaymentResult(false, null, latestResult?.FailureReason, maxAttempts);
    }

    private OutboxMessage CreateOutboxMessage<TPayload>(
        long checkoutAttemptId,
        OutboxMessageType type,
        TPayload payload)
    {
        return new OutboxMessage
        {
            CheckoutAttemptId = checkoutAttemptId,
            Type = type,
            Status = OutboxStatus.Pending,
            PayloadJson = JsonSerializer.Serialize(payload, JsonOptions),
            CreatedAt = clock.UtcNow
        };
    }

    private static CheckoutResponse ToResponse(CheckoutAttempt attempt)
    {
        var integrations = attempt.OutboxMessages
            .OrderBy(message => message.Type)
            .Select(message => new IntegrationStatusDto(
                message.Type,
                message.Status,
                message.AttemptCount,
                message.LastError))
            .ToArray();

        return new CheckoutResponse(
            attempt.Id,
            attempt.OrderId,
            attempt.Status,
            attempt.PaymentTransaction?.Status,
            attempt.FailureReason ?? attempt.PaymentTransaction?.FailureReason,
            integrations);
    }

    private sealed record RetriedPaymentResult(
        bool Succeeded,
        string? ProviderTransactionId,
        string? FailureReason,
        int AttemptCount);
}

File: src/main/java/com/example/managementplatform/CheckoutService.java
package com.example.managementplatform;

import java.time.Instant;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;

class CheckoutService {
    private final ManagementRepository repository;

    CheckoutService(ManagementRepository repository) {
        this.repository = repository;
    }

    synchronized CheckoutResponse checkout(long orderId, CheckoutRequest request) {
        var idempotencyKey = requireText(request.idempotencyKey(), "An idempotency key is required.");
        var token = requireText(request.paymentMethodToken(), "A payment method token is required.");
        return repository.findCheckout(orderId, idempotencyKey).map(this::toResponse).orElseGet(() -> createCheckout(orderId, idempotencyKey, token));
    }

    CheckoutResponse get(long checkoutId) {
        return repository.findCheckout(checkoutId).map(this::toResponse)
            .orElseThrow(() -> new ApiException(404, "Checkout was not found."));
    }

    private CheckoutResponse createCheckout(long orderId, String idempotencyKey, String token) {
        var order = repository.findOrder(orderId)
            .orElseThrow(() -> new ApiException(404, "Order was not found."));
        if (order.status() != OrderStatus.DRAFT) {
            throw new ApiException(409, "Order is already paid or being processed.");
        }

        order.markProcessing();
        var attempt = repository.createCheckout(order.id(), idempotencyKey);
        var payment = charge(order, attempt, token);
        attempt.paymentTransaction(payment);

        if (payment.status() == PaymentStatus.FAILED) {
            var reason = payment.failureReason() == null ? "Payment was declined." : payment.failureReason();
            attempt.fail(reason, Instant.now());
            order.markDraft();
            attempt.outboxMessages().add(new OutboxMessage(attempt.id(), OutboxMessageType.PAYMENT_CHARGE, OutboxStatus.FAILED,
                "{\"orderId\":" + order.id() + ",\"checkoutAttemptId\":" + attempt.id() + "}", payment.attemptCount(), reason, Instant.now()));
            return toResponse(attempt);
        }

        attempt.succeed(Instant.now());
        order.markPaid(Instant.now());
        attempt.outboxMessages().add(new OutboxMessage(attempt.id(), OutboxMessageType.SEND_CHECKOUT_EMAIL, OutboxStatus.PENDING,
            "{\"orderId\":" + order.id() + ",\"email\":\"" + order.tenant().email() + "\"}", 0, null, Instant.now()));
        attempt.outboxMessages().add(new OutboxMessage(attempt.id(), OutboxMessageType.PUSH_TO_PRODUCTION, OutboxStatus.PENDING,
            "{\"orderId\":" + order.id() + ",\"name\":\"" + order.name() + "\"}", 0, null, Instant.now()));
        return toResponse(attempt);
    }

    private PaymentTransaction charge(Order order, CheckoutAttempt attempt, String token) {
        var normalized = token.toLowerCase();
        if (normalized.contains("fail") || normalized.contains("decline")) {
            return new PaymentTransaction(attempt.id(), PaymentStatus.FAILED, 1, order.amount(), order.currency(), null,
                "Mock payment gateway declined the payment.", Instant.now());
        }
        return new PaymentTransaction(attempt.id(), PaymentStatus.SUCCEEDED, normalized.contains("retry") ? 2 : 1,
            order.amount(), order.currency(), "txn_" + UUID.randomUUID(), null, Instant.now());
    }

    private String requireText(String value, String message) {
        if (value == null || value.isBlank()) {
            throw new ApiException(400, message);
        }
        return value.trim();
    }

    CheckoutResponse toResponse(CheckoutAttempt attempt) {
        var integrations = attempt.outboxMessages().stream()
            .sorted(Comparator.comparing(OutboxMessage::type))
            .map(message -> new IntegrationStatusDto(message.type(), message.status(), message.attemptCount(), message.lastError()))
            .toList();
        var payment = attempt.paymentTransaction();
        return new CheckoutResponse(attempt.id(), attempt.orderId(), attempt.status(), payment == null ? null : payment.status(),
            attempt.failureReason() == null && payment != null ? payment.failureReason() : attempt.failureReason(), integrations);
    }
}
